packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts

```diff
@@ -55,6 +70,85 @@ export class DataObjectFactory<
 			mergedObjects.push(SharedMap.getFactory());
 		}
 
-		super(type, ctor, mergedObjects, optionalProviders, registryEntries, runtimeFactory);
+		let converted = false;
+		const convertRoundTripP = new Deferred<void>();
+
+		const fullConvertDataFn =
+			convertDataFn === undefined
+				? undefined
+				: async (runtime: FluidDataStoreRuntime) => {
+						if (!converted) {
+							converted = true;
+							submitConversionOp(runtime);
+							await convertRoundTripP.promise;
+
+							const root = (await runtime.getChannel(
```

Comment:
can we leave loading of DDSs to converter function? While root directly is somewhat special, I'd want to make it less special (for converter logic), as our initial concern is to convert legacy tree DDS (and ideally - all DDSs to single instance of the tree). I'm not sure how much it helps to load a root instance here. Plus, long term, I think we should find a way to daisy chain converters and allow multiple conversions (so there could be no longer Directory instance here when some converter runs).